README.md

# localstorage-retry
[![Circle CI](https://circleci.com/gh/segmentio/localstorage-retry.svg?style=shield&circle-token=26daea4c3c8e5645f15841fdda51f14386bc5302)](https://circleci.com/gh/segmentio/localstorage-retry)

Provides durable retries with a queue held in `localStorage` (with graceful fallbacks to memory when necessary).

## How It Works

Each page maintains its own list of queued and in-progress tasks, while constantly refreshing its ack time. If a queue goes more than 10s without updating its ack, another page will remove it, claim all queued tasks, and retry all in-progress tasks.

## API

### new Queue(name, [opts], processFunc(item, done(err, res)))

You can omit the `opts` argument to initialize the queue with defaults:

```javascript
var Queue = require('@segment/localstorage-retry');

var queue = new Queue('my_queue_name', function process(item, done) {
  sendAsync(item, function(err, res) {
    if (err) return done(err);
    done(null, res);
  });
});

queue.on('processed', function(err, res, item) {
  if (err) return console.warn('processing %O failed with error %O', item, err);
  console.log('successfully sent %O with response %O', item, res);
});

queue.start();
```

### Options

The queue can be initialized with the following options (*defaults shown*):

```js
var options = {
  minRetryDelay: 1000,   // min retry delay in ms (used in exp. backoff calcs)
  maxRetryDelay: 30000,  // max retry delay in ms (used in exp. backoff calcs)
  backoffFactor: 2,      // exponential backoff factor (attempts^n)
  backoffJitter: 0,      // jitter factor for backoff calcs (0 is usually fine)
  maxItems: Infinity     // queue high water mark (we suggest 100 as a max)
  maxAttempts: Infinity  // max retry attempts before discarding
};

var queue = new Queue('my_queue_name', options, (item, done) => {
  sendAsync(item, (err, res) => {
    if (err) return done(err);
    done(null, res);
  });
});

queue.start();
```

### .addItem(item)

Adds an item to the queue

```javascript
queue.addItem({ a: 'b' });
```

### .getDelay `(attemptNumber) -> ms`

Can be overridden to provide a custom retry delay in ms. You'll likely want to use the queue instance's backoff constants here.

```js
this.backoff = {
  MIN_RETRY_DELAY: opts.minRetryDelay || 1000,
  MAX_RETRY_DELAY: opts.maxRetryDelay || 30000,
  FACTOR: opts.backoffFactor || 2,
  JITTER: opts.backoffJitter || 0
};
```

Default implementation:

```javascript
queue.getDelay = function(attemptNumber) {
  var ms = this.backoff.MIN_RETRY_DELAY * Math.pow(this.backoff.FACTOR, attemptNumber);
  if (this.backoff.JITTER) {
    var rand =  Math.random();
    var deviation = Math.floor(rand * this.backoff.JITTER * ms);
    if (Math.floor(rand * 10) < 5) {
      ms -= deviation;
    } else {
      ms += deviation;
    }
  }
  return Number(Math.min(ms, this.backoff.MAX_RETRY_DELAY).toPrecision(1));
};
```

### .shouldRetry `(item, attemptNumber, error) -> boolean`

Can be overridden to provide custom logic for whether to requeue the item. You'll likely want to use the queue instance's `maxAttempts` variable (which is overridable via constructor's `opts` argument).

**Default**:
```js
queue.shouldRetry = function(item, attemptNumber, error) {
  if (attemptNumber > this.maxAttempts) return false;
  return true;
};
```

You may also want to selectively retry based on error returned by your process function or something in the item itself.

**Override Example**:
```javascript
queue.shouldRetry = function(item, attemptNumber, error) {
  // max attempts
  if (attemptNumber > this.maxAttempts) return false;

  // based on something in the item itself
  if (new Date(item.timestamp) - new Date() > 86400000) return false;

  // selective error handling
  if (error.code === '429') return false;

  return true;
}
```

### .start

Starts the queue processing items. Anything added before calling `.start` will be queued until `.start` is called.

```javascript
queue.start();
```

### .stop

Stops the queue from processing. Any retries queued may be picked claimed by another queue after a timeout.

```javascript
queue.stop();
```

## Emitter

You can listen for `processed` events, which are emitted with each invocation of the `processFunc` and passed any error or response provided along with the item itself.

If a message is discarded entirely because it does not pass your `shouldRetry` logic upon attempted re-enqueuing, the queue will emit a `discard` event.

### `processed`

```javascript
queue.on('processed', function(err, res, item) {
  if (err) return console.warn('processing %O failed with error %O', item, err);
  console.log('successfully sent %O with response %O', item, res);
});
```

### `discard`

```javascript
queue.on('discard', function(item, attempts, error) {
  console.error('discarding message %O after %d attempts with last known error %O', item, attempts, error);
})
```

## License

Released under the [MIT License](LICENSE)


HISTORY.md
# 1.2.4 / 2020-08-11

- [Fix](https://github.com/segmentio/localstorage-retry/pull/21): Modified reclaim to run on original engine with tests 

# 1.2.3 / 2020-05-10

- [Fix](https://github.com/segmentio/localstorage-retry/pull/17): Re order keys to delete ack last

# 1.2.2 / 2018-07-19

- [Fix](https://github.com/segmentio/localstorage-retry/pull/13): Respect attempts when reclaiming queue
- [Fix](https://github.com/segmentio/localstorage-retry/pull/11): Limit inProgress using maxItems

# 1.2.1 / 2018-07-11

- [Fix](https://github.com/segmentio/localstorage-retry/pull/10): Fix object loss when adding multiple items to the queue.

# 1.2.0 / 2017-08-23

- add support for configurable retry backoffs
- add maxItems option to constrain max queue size
- add maxAttempts option to constrain number of attempts
- update default shouldretry and getdelay logic to take advantage of HWM and exp. backoff with jitter

# 1.1.0 / 2017-08-11

- Minor Refactor
- Fallback to in-memory storage engine in the event of exceeding localStorage quota

# 1.0.3 / 2017-08-02

- Default _all_ localstorage queue retrievals to empty data structures

# 1.0.2 / 2017-08-02

- Default to empty queue

# 1.0.1 / 2017-07-12

- Bump package.json

# 1.0.0 / 2017-07-12

- Retry Logic and Events (#3)

# 0.0.2 / 2017-01-23

- Fix setTimeout invocation

# 0.0.1 / 2017-01-09

- Initial Implementation


test/index.test.js
'use strict';

var Queue = require('../');
var Store = require('../lib/store');
var engine = require('../lib/engine').defaultEngine;
var sinon = require('sinon');
var lolex = require('lolex');
var assert = require('proclaim');
var Schedule = require('../lib/schedule');

describe('Queue', function() {
  var queue;
  var clock;

  beforeEach(function() {
    engine.clear();
    clock = lolex.createClock(0);
    Schedule.setClock(clock);

    // Have the default function be a spied success
    queue = new Queue('test', sinon.spy(function(_, cb) {
      return cb();
    }));
  });

  afterEach(function() {
    queue.stop();
    Schedule.resetClock();
  });

  it('should run a task', function() {
    queue.start();

    queue.addItem('a');

    assert(queue.fn.calledOnce);
    assert(queue.fn.calledWith('a'));
  });

  it('should retry a task if it fails', function() {
    queue.start();

    // Fail the first time
    queue.fn = sinon.spy(function(_, cb) {
      return cb(new Error('no'));
    });
    queue.addItem('a');
    assert(queue.fn.calledOnce);
    assert(queue.fn.calledWith('a'));

    // Succeed the second time
    queue.fn = sinon.spy(function(_, cb) {
      return cb();
    });

    // Delay for the first retry
    clock.tick(queue.getDelay(1));
    assert(queue.fn.calledOnce);
    assert(queue.fn.calledWith('a'));
  });

  it('should delay retries', function() {
    queue.start();

    queue.requeue('b', 1);
    queue.addItem('a');

    assert(queue.fn.calledOnce);
    assert(queue.fn.calledWith('a'));

    queue.fn.reset();

    // delay for the retry
    clock.tick(queue.getDelay(1));

    assert(queue.fn.calledOnce);
    assert(queue.fn.calledWith('b'));
  });

  it('should respect shouldRetry', function() {
    queue.shouldRetry = function(_, attemptNumber) {
      if (attemptNumber > 2) return false;
      return true;
    };
    // Fail
    queue.fn = sinon.spy(function(_, cb) {
      return cb(new Error('no'));
    });
    queue.start();

    // over maxattempts
    queue.requeue('a', 3);
    clock.tick(queue.getDelay(3));
    assert(queue.fn.notCalled);

    queue.fn.reset();

    queue.requeue('a', 2);
    clock.tick(queue.getDelay(2));
    assert(queue.fn.calledOnce);

    // logic based on item state (eg. could be msg timestamp field)
    queue.shouldRetry = function(item) {
      if (item.shouldRetry === false) return false;
      return true;
    };

    queue.fn.reset();
    queue.requeue({ shouldRetry: false }, 1);
    clock.tick(queue.getDelay(1));
    assert(queue.fn.notCalled);
  });

  it('should respect maxItems', function() {
    queue.maxItems = 100;

    for (var i = 0; i < 105; i++) {
      clock.tick(1);
      queue.addItem(i);
    }

    var _queue = queue._store.get(queue.keys.QUEUE);
    assert.equal(_queue.length, 100);
    assert.equal(_queue[0].item, 5);
    assert.equal(_queue[99].item, 104);
  });

  it('should take over a queued task if a queue is abandoned', function() {
    // a wild queue of interest appears
    var foundQueue = new Store('test', 'fake-id', queue.keys);
    foundQueue.set(foundQueue.keys.ACK, 0); // fake timers starts at time 0
    foundQueue.set(foundQueue.keys.QUEUE, [{
      item: 'a',
      time: 0,
      attemptNumber: 0
    }]);

    // wait for the queue to expire
    clock.tick(queue.timeouts.RECLAIM_TIMEOUT);

    queue.start();

    // wait long enough for the other queue to expire and be reclaimed
    clock.tick(
      queue.timeouts.RECLAIM_TIMER
      + queue.timeouts.RECLAIM_WAIT * 2
    );

    assert(queue.fn.calledOnce);
    assert(queue.fn.calledWith('a'));
  });

  it('should take over an in-progress task if a queue is abandoned', function() {
    // set up a fake queue
    var foundQueue = new Store('test', 'fake-id', queue.keys);
    foundQueue.set(foundQueue.keys.ACK, -15000);
    foundQueue.set(foundQueue.keys.IN_PROGRESS, {
      'task-id': {
        item: 'a',
        time: 0,
        attemptNumber: 0
      }
    });

    // wait for the queue to expire
    clock.tick(queue.timeouts.RECLAIM_TIMEOUT);

    queue.start();

    // wait long enough for the other queue to expire and be reclaimed
    clock.tick(
      queue.timeouts.RECLAIM_TIMER
      + queue.timeouts.RECLAIM_WAIT * 2
    );

    assert(queue.fn.calledOnce);
    assert(queue.fn.calledWith('a'));
  });

  it('should take over multiple tasks if a queue is abandoned', function() {
    // set up a fake queue
    var foundQueue = new Store('test', 'fake-id', queue.keys);
    foundQueue.set(foundQueue.keys.ACK, -15000);
    foundQueue.set(foundQueue.keys.QUEUE, [{
      item: 'a',
      time: 0,
      attemptNumber: 0
    }]);
    foundQueue.set(foundQueue.keys.IN_PROGRESS, {
      'task-id': {
        item: 'b',
        time: 1,
        attemptNumber: 0
      }
    });

    // wait for the queue to expire
    clock.tick(queue.timeouts.RECLAIM_TIMEOUT);

    queue.start();

    // wait long enough for the other queue to expire and be reclaimed
    clock.tick(
      queue.timeouts.RECLAIM_TIMER
      + queue.timeouts.RECLAIM_WAIT * 2
    );

    assert(queue.fn.calledTwice);
    assert(queue.fn.calledWith('a'));
    assert(queue.fn.calledWith('b'));
  });

  describe('while using in memory engine', function() {
    beforeEach(function() {
      queue._store._swapEngine();
    });

    it('should take over a queued task if a queue is abandoned', function() {
      // a wild queue of interest appears
      var foundQueue = new Store('test', 'fake-id', queue.keys);
      foundQueue.set(foundQueue.keys.ACK, 0); // fake timers starts at time 0
      foundQueue.set(foundQueue.keys.QUEUE, [{
        item: 'a',
        time: 0,
        attemptNumber: 0
      }]);
  
      // wait for the queue to expire
      clock.tick(queue.timeouts.RECLAIM_TIMEOUT);
  
      queue.start();
  
      // wait long enough for the other queue to expire and be reclaimed
      clock.tick(
        queue.timeouts.RECLAIM_TIMER
        + queue.timeouts.RECLAIM_WAIT * 2
      );
  
      assert(queue.fn.calledOnce);
      assert(queue.fn.calledWith('a'));
    });
  
    it('should take over an in-progress task if a queue is abandoned', function() {
      // set up a fake queue
      var foundQueue = new Store('test', 'fake-id', queue.keys);
      foundQueue.set(foundQueue.keys.ACK, -15000);
      foundQueue.set(foundQueue.keys.IN_PROGRESS, {
        'task-id': {
          item: 'a',
          time: 0,
          attemptNumber: 0
        }
      });
  
      // wait for the queue to expire
      clock.tick(queue.timeouts.RECLAIM_TIMEOUT);
  
      queue.start();
  
      // wait long enough for the other queue to expire and be reclaimed
      clock.tick(
        queue.timeouts.RECLAIM_TIMER
        + queue.timeouts.RECLAIM_WAIT * 2
      );
  
      assert(queue.fn.calledOnce);
      assert(queue.fn.calledWith('a'));
    });
  
    it('should take over multiple tasks if a queue is abandoned', function() {
      // set up a fake queue
      var foundQueue = new Store('test', 'fake-id', queue.keys);
      foundQueue.set(foundQueue.keys.ACK, -15000);
      foundQueue.set(foundQueue.keys.QUEUE, [{
        item: 'a',
        time: 0,
        attemptNumber: 0
      }]);
      foundQueue.set(foundQueue.keys.IN_PROGRESS, {
        'task-id': {
          item: 'b',
          time: 1,
          attemptNumber: 0
        }
      });
  
      // wait for the queue to expire
      clock.tick(queue.timeouts.RECLAIM_TIMEOUT);
  
      queue.start();
  
      // wait long enough for the other queue to expire and be reclaimed
      clock.tick(
        queue.timeouts.RECLAIM_TIMER
        + queue.timeouts.RECLAIM_WAIT * 2
      );
  
      assert(queue.fn.calledTwice);
      assert(queue.fn.calledWith('a'));
      assert(queue.fn.calledWith('b'));
    });
  });

  it('should respect maxAttempts when rejected', function() {
    var calls = new Array(100);

    queue.maxItems = calls.length;
    queue.maxAttempts = 2;

    queue.fn = function(item, done) {
      if (!calls[item.index]) {
        calls[item.index] = 1;
      } else {
        calls[item.index]++;
      }

      done(new Error());
    };

    for (var i = 0; i < calls.length; i++) {
      queue.addItem({ index: i });
    }

    queue.start();

    clock.tick(queue.getDelay(1) + queue.getDelay(2));
    calls.forEach(function(call) {
      assert(call === queue.maxAttempts + 1);
    });
  });

  it('should limit inProgress using maxItems', function() {
    var waiting = [];
    var i;

    queue.maxItems = 100;
    queue.maxAttempts = 2;
    queue.fn = function(_, done) {
      waiting.push(done);
    };

    // add maxItems * 2 items
    for (i = 0; i < queue.maxItems * 2; i++) {
      queue.addItem({ index: i });
    }
    
    // the queue should be full
    assert(size(queue).queue === queue.maxItems);

    queue.start();
    // the queue is now empty and everything is in progress
    assert(size(queue).queue === 0);
    assert(size(queue).inProgress === queue.maxItems);

    // while the items are in progress let's add maxItems times two items
    for (i = 0; i < queue.maxItems * 2; i++) {
      queue.addItem({ index: i });
    }

    // inProgress and queue should be full
    assert(size(queue).queue === queue.maxItems);
    assert(size(queue).inProgress === queue.maxItems);
    assert(waiting.length === queue.maxItems);

    // resolved all waiting items
    while (waiting.length) {
      waiting.pop()();
    }

    // inProgress should now be empty
    assert(size(queue).queue === queue.maxItems);
    assert(size(queue).inProgress === 0);

    // wait for the queue to be processed
    clock.tick(queue.getDelay(0));

    // items should now be in progress
    assert(size(queue).queue === 0);
    assert(size(queue).inProgress === queue.maxItems);

    function size(queue) {
      return {
        queue: queue._store.get(queue.keys.QUEUE).length,
        inProgress: Object.keys(queue._store.get(queue.keys.IN_PROGRESS) || {}).length
      };
    }
  });
});

describe('events', function() {
  var queue;
  beforeEach(function() {
    queue = new Queue('events', function(_, cb) {
      cb();
    });
  });

  afterEach(function() {
    queue.stop();
  });

  it('should emit processed with response, and item', function(done) {
    queue.fn = function(item, cb) {
      cb(null, { text: 'ok' });
    };
    queue.on('processed', function(err, res, item) {
      if (err) done(err);
      assert.equal(item.a, 'b');
      assert.equal(res.text, 'ok');
      done();
    });
    queue.start();
    queue.addItem({ a: 'b' });
  });

  it('should include errors in callback to processed event', function(done) {
    queue.fn = function(item, cb) {
      cb(new Error('fail'));
    };

    queue.on('processed', function(err, res, item) {
      assert.equal(item.a, 'c');
      assert.equal(err && err.message, 'fail');
      done();
    });

    queue.start();
    queue.addItem({ a: 'c' });
  });

  it('should emit discard if the message fails shouldRetry', function(done) {
    queue.fn = function(item, cb) {
      cb(new Error('no'));
    };
    queue.shouldRetry = function(item, attemptNumber) {
      return attemptNumber < 2;
    };
    queue.on('discard', function(item, attempts) {
      assert.equal(item.a, 'b');
      assert.equal(attempts, 2);
      done();
    });
    queue.start();
    queue.addItem({ a: 'b' });
  });
});

describe('end-to-end', function() {
  var queue;
  beforeEach(function() {
    queue = new Queue('e2e_test', function(_, cb) { cb(); });
  });

  afterEach(function() {
    queue.stop();
  });

  it('should run end-to-end', function(done) {
    queue.fn = function(item, cb) {
      cb();
      done();
    };
    queue.start();
    queue.addItem({ a: 'b' });
  });

  it('should run end-to-end async', function(done) {
    queue.fn = function(item, cb) {
      setTimeout(function() {
        cb();
      }, 1000);
    };
    queue.on('processed', function() {
      done();
    });

    queue.start();
    queue.addItem({ a: 'b' });
  });
});


test/store.test.js
'use strict';

var Store = require('../lib/store');
var engine = require('../lib/engine').defaultEngine;
var inMemoryEngine = require('../lib/engine').inMemoryEngine;
var assert = require('proclaim');
var each = require('@ndhoule/each');

describe('Store', function() {
  var store;
  var keys = {
    IN_PROGRESS: 'inProgress',
    QUEUE: 'queue',
    ACK: 'ack',
    RECLAIM_START: 'reclaimStart',
    RECLAIM_END: 'reclaimEnd'
  };

  beforeEach(function() {
    engine.clear();
    store = new Store('name', 'id', keys);
  });

  describe('.get', function() {
    it('should default to null', function() {
      each(function(k) {
        assert.strictEqual(store.get(k), null);
      }, keys);
    });

    it('should de-serialize json', function() {
      each(function(k) {
        engine.setItem('name.id.' + k, '["a","b",{}]');
        assert.deepEqual(store.get(k), [ 'a', 'b', {} ]);
      }, keys);
    });

    it('should return null if value is not valid json', function() {
      engine.setItem('name.id.queue', '[{]}');
      assert.strictEqual(store.get(keys.QUEUE), null);
    });
  });

  describe('.set', function() {
    it('should serialize json', function() {
      each(function(k) {
        store.set(k, ['a', 'b', {}]);
        assert.strictEqual(engine.getItem('name.id.' + k), '["a","b",{}]');
      }, keys);
    });
  });

  describe('.remove', function() {
    it('should remove the item', function() {
      each(function(k) {
        store.set(k, 'a');
        store.remove(k);
        assert.strictEqual(engine.getItem('name.id.' + k), null);
      }, keys);
    });
  });

  describe('._createValidKey', function() {
    it('should return compound if no keys specd', function() {
      store = new Store('name', 'id');
      assert.strictEqual(store._createValidKey('test'), 'name.id.test');
    });

    it('should return undefined if invalid key', function() {
      store = new Store('name', 'id', { nope: 'wrongKey' });
      assert.strictEqual(store._createValidKey('test'), undefined);
    });

    it('should return compound if valid key', function() {
      assert.strictEqual(store._createValidKey('queue'), 'name.id.queue');
    });
  });

  // Skip the "swap" tests when not applicable
  var xdescribe = window.localStorage ? describe : describe.skip;

  xdescribe('._swapEngine', function() {
    it('should switch the underlying storage mechanism', function() {
      assert.strictEqual(store.engine, engine);
      store._swapEngine();
      assert.strictEqual(store.engine, inMemoryEngine);
    });

    it('should not switch the original storage mechanism', function() {
      assert.strictEqual(store.getOriginalEngine(), engine);
      store._swapEngine();
      assert.strictEqual(store.getOriginalEngine(), engine);
    });

    it('should swap upon quotaExceeded on set', function() {
      var lsProxy =  {
        length: window.localStorage.length,
        setItem: function(k, v) {
          return window.localStorage.setItem(k, v);
        },
        getItem: function(k) {
          return window.localStorage.getItem(k);
        },
        removeItem: function(k) {
          return window.localStorage.removeItem(k);
        },
        clear: function() {
          return window.localStorage.clear();
        },
        key: function(i) {
          return window.localStorage.key(i);
        }
      };

      store = new Store('name', 'id', keys, lsProxy);

      each(function(v) {
        store.set(v, 'stuff');
      }, keys);
      store.engine.setItem = function() {
        // eslint-disable-next-line no-throw-literal
        throw { code: 22, number: -2147024882 };
      };
      store.set(keys.QUEUE, 'other');
      assert.strictEqual(store.get(keys.QUEUE), 'other');
    });
  });
});


test/engine.test.js
'use strict';

var defaultEngine = require('../lib/engine').defaultEngine;
var inMemoryEngine = require('../lib/engine').inMemoryEngine;
var assert = require('proclaim');

// Skip the "supported" tests when its not actually supported
var xdescribe = window.localStorage ? describe : describe.skip;

describe('localStorage', function() {
  var engine;

  xdescribe('when supported', function() {
    beforeEach(function() {
      engine = defaultEngine;
      engine.clear();
    });

    it('should function', function() {
      engine.setItem('test-key', 'abc');
      assert.strictEqual(engine.getItem('test-key'), 'abc');
      assert.strictEqual(engine.length, 1);
      assert.strictEqual(engine.key(0), 'test-key');

      engine.removeItem('test-key');
      assert.strictEqual(engine.getItem('test-key'), null);
      assert.strictEqual(engine.length, 0);

      engine.setItem('test-key', 'abc');
      engine.clear();
      assert.strictEqual(engine.length, 0);
    });
  });

  describe('when not supported', function() {
    beforeEach(function() {
      engine = inMemoryEngine;
      engine.clear();
    });

    it('should function', function() {
      engine.setItem('test-key', 'abc');
      assert.strictEqual(engine.getItem('test-key'), 'abc');
      assert.strictEqual(engine.length, 1);
      assert.strictEqual(engine.key(0), 'test-key');

      engine.removeItem('test-key');
      assert.strictEqual(engine.getItem('test-key'), null);
      assert.strictEqual(engine.length, 0);

      engine.setItem('test-key', 'abc');
      engine.clear();
      assert.strictEqual(engine.length, 0);
    });
  });
});


lib/engine.js
'use strict';

var keys = require('@ndhoule/keys');
var uuid = require('uuid').v4;

var inMemoryStore = {
  _data: {},
  length: 0,
  setItem: function(key, value) {
    this._data[key] = value;
    this.length = keys(this._data).length;
    return value;
  },
  getItem: function(key) {
    if (key in this._data) {
      return this._data[key];
    }
    return null;
  },
  removeItem: function(key) {
    if (key in this._data) {
      delete this._data[key];
    }
    this.length = keys(this._data).length;
    return null;
  },
  clear: function() {
    this._data = {};
    this.length = 0;
  },
  key: function(index) {
    return keys(this._data)[index];
  }
};

function isSupportedNatively() {
  try {
    if (!window.localStorage) return false;
    var key = uuid();
    window.localStorage.setItem(key, 'test_value');
    var value = window.localStorage.getItem(key);
    window.localStorage.removeItem(key);

    // handle localStorage silently failing
    return value === 'test_value';
  } catch (e) {
    // Can throw if localStorage is disabled
    return false;
  }
}

function pickStorage() {
  if (isSupportedNatively()) {
    return window.localStorage;
  }
  // fall back to in-memory
  return inMemoryStore;
}

// Return a shared instance
module.exports.default = pickStorage();
// Expose the in-memory store explicitly for testing
module.exports.inMemoryStore = inMemoryStore;


lib/index.js
'use strict';

var uuid = require('uuid').v4;
var Store = require('./store');
var each = require('@ndhoule/each');
var Schedule = require('./schedule');
var debug = require('debug')('localstorage-retry');
var Emitter = require('component-emitter');

// Some browsers don't support Function.prototype.bind, so just including a simplified version here
function bind(func, obj) {
  return function() {
    return func.apply(obj, arguments);
  };
}

/**
 * @callback processFunc
 * @param {Mixed} item The item added to the queue to process
 * @param {Function} done A function to call when processing is completed.
 *   @param {Error} Optional error parameter if the processing failed
 *   @param {Response} Optional response parameter to emit for async handling
 */

/**
 * Constructs a Queue backed by localStorage
 *
 * @constructor
 * @param {String} name The name of the queue. Will be used to find abandoned queues and retry their items
 * @param {processFunc} fn The function to call in order to process an item added to the queue
 */
function Queue(name, opts, fn) {
  if (typeof opts === 'function') fn = opts;
  this.name = name;
  this.id = uuid();
  this.fn = fn;
  this.maxItems = opts.maxItems || Infinity;
  this.maxAttempts = opts.maxAttempts || Infinity;

  this.backoff = {
    MIN_RETRY_DELAY: opts.minRetryDelay || 1000,
    MAX_RETRY_DELAY: opts.maxRetryDelay || 30000,
    FACTOR: opts.backoffFactor || 2,
    JITTER: opts.backoffJitter || 0
  };

  // painstakingly tuned. that's why they're not "easily" configurable
  this.timeouts = {
    ACK_TIMER: 1000,
    RECLAIM_TIMER: 3000,
    RECLAIM_TIMEOUT: 10000,
    RECLAIM_WAIT: 500
  };

  this.keys = {
    IN_PROGRESS: 'inProgress',
    QUEUE: 'queue',
    RECLAIM_START: 'reclaimStart',
    RECLAIM_END: 'reclaimEnd',
    ACK: 'ack'
  };

  this._schedule = new Schedule();
  this._processId = 0;

  // Set up our empty queues
  this._store = new Store(this.name, this.id, this.keys);
  this._store.set(this.keys.IN_PROGRESS, {});
  this._store.set(this.keys.QUEUE, []);

  // bind recurring tasks for ease of use
  this._ack = bind(this._ack, this);
  this._checkReclaim = bind(this._checkReclaim, this);
  this._processHead = bind(this._processHead, this);

  this._running = false;
}

/**
 * Mix in event emitter
 */

Emitter(Queue.prototype);

/**
 * Starts processing the queue
 */
Queue.prototype.start = function() {
  if (this._running) {
    this.stop();
  }
  this._running = true;
  this._ack();
  this._checkReclaim();
  this._processHead();
};

/**
 * Stops processing the queue
 */
Queue.prototype.stop = function() {
  this._schedule.cancelAll();
  this._running = false;
};

/**
 * Decides whether to retry. Overridable.
 *
 * @param {Object} item The item being processed
 * @param {Number} attemptNumber The attemptNumber (1 for first retry)
 * @param {Error} error The error from previous attempt, if there was one
 * @return {Boolean} Whether to requeue the message
 */
Queue.prototype.shouldRetry = function(_, attemptNumber) {
  if (attemptNumber > this.maxAttempts) return false;
  return true;
};

/**
 * Calculates the delay (in ms) for a retry attempt
 *
 * @param {Number} attemptNumber The attemptNumber (1 for first retry)
 * @return {Number} The delay in milliseconds to wait before attempting a retry
 */
Queue.prototype.getDelay = function(attemptNumber) {
  var ms = this.backoff.MIN_RETRY_DELAY * Math.pow(this.backoff.FACTOR, attemptNumber);
  if (this.backoff.JITTER) {
    var rand =  Math.random();
    var deviation = Math.floor(rand * this.backoff.JITTER * ms);
    if (Math.floor(rand * 10) < 5) {
      ms -= deviation;
    } else {
      ms += deviation;
    }
  }
  return Number(Math.min(ms, this.backoff.MAX_RETRY_DELAY).toPrecision(1));
};

/**
 * Adds an item to the queue
 *
 * @param {Mixed} item The item to process
 */
Queue.prototype.addItem = function(item) {
  this._enqueue({
    item: item,
    attemptNumber: 0,
    time: this._schedule.now()
  });
};

/**
 * Adds an item to the retry queue
 *
 * @param {Mixed} item The item to retry
 * @param {Number} attemptNumber The attempt number (1 for first retry)
 * @param {Error} [error] The error from previous attempt, if there was one
 */
Queue.prototype.requeue = function(item, attemptNumber, error) {
  if (this.shouldRetry(item, attemptNumber, error)) {
    this._enqueue({
      item: item,
      attemptNumber: attemptNumber,
      time: this._schedule.now() + this.getDelay(attemptNumber)
    });
  } else {
    this.emit('discard', item, attemptNumber);
  }
};

Queue.prototype._enqueue = function(entry) {
  var queue = this._store.get(this.keys.QUEUE) || [];
  queue = queue.slice(-(this.maxItems - 1));
  queue.push(entry);
  queue = queue.sort(function(a,b) {
    return a.time - b.time;
  });

  this._store.set(this.keys.QUEUE, queue);

  if (this._running) {
    this._processHead();
  }
};

Queue.prototype._processHead = function() {
  var self = this;
  var store = this._store;

  // cancel the scheduled task if it exists
  this._schedule.cancel(this._processId);

  // Pop the head off the queue
  var queue = store.get(this.keys.QUEUE) || [];
  var inProgress = store.get(this.keys.IN_PROGRESS) || {};
  var now = this._schedule.now();
  var toRun = [];

  function enqueue(el, id) {
    toRun.push({
      item: el.item,
      done: function handle(err, res) {
        var inProgress = store.get(self.keys.IN_PROGRESS) || {};
        delete inProgress[id];
        store.set(self.keys.IN_PROGRESS, inProgress);
        self.emit('processed', err, res, el.item);
        if (err) {
          self.requeue(el.item, el.attemptNumber + 1, err);
        }
      }
    });
  }

  var inProgressSize = Object.keys(inProgress).length;

  while (queue.length && queue[0].time <= now && inProgressSize++ < self.maxItems) {
    var el = queue.shift();
    var id = uuid();

    // Save this to the in progress map
    inProgress[id] = {
      item: el.item,
      attemptNumber: el.attemptNumber,
      time: self._schedule.now()
    };

    enqueue(el, id);
  }

  store.set(this.keys.QUEUE, queue);
  store.set(this.keys.IN_PROGRESS, inProgress);

  each(function(el) {
    // TODO: handle fn timeout
    try {
      self.fn(el.item, el.done);
    } catch (err) {
      debug('Process function threw error: ' + err);
    }
  }, toRun);

  // re-read the queue in case the process function finished immediately or added another item
  queue = store.get(this.keys.QUEUE) || [];
  this._schedule.cancel(this._processId);
  if (queue.length > 0) {
    this._processId = this._schedule.run(this._processHead, queue[0].time - now);
  }
};

// Ack continuously to prevent other tabs from claiming our queue
Queue.prototype._ack = function() {
  this._store.set(this.keys.ACK, this._schedule.now());
  this._store.set(this.keys.RECLAIM_START, null);
  this._store.set(this.keys.RECLAIM_END, null);
  this._schedule.run(this._ack, this.timeouts.ACK_TIMER);
};

Queue.prototype._checkReclaim = function() {
  var self = this;

  function tryReclaim(store) {
    store.set(self.keys.RECLAIM_START, self.id);
    store.set(self.keys.ACK, self._schedule.now());

    self._schedule.run(function() {
      if (store.get(self.keys.RECLAIM_START) !== self.id) return;
      store.set(self.keys.RECLAIM_END, self.id);

      self._schedule.run(function() {
        if (store.get(self.keys.RECLAIM_END) !== self.id) return;
        if (store.get(self.keys.RECLAIM_START) !== self.id) return;
        self._reclaim(store.id);
      }, self.timeouts.RECLAIM_WAIT);
    }, self.timeouts.RECLAIM_WAIT);
  }

  function findOtherQueues(name) {
    var res = [];
    var storage = self._store.getOriginalEngine();
    for (var i = 0; i < storage.length; i++) {
      var k = storage.key(i);
      var parts = k.split('.');
      if (parts.length !== 3) continue;
      if (parts[0] !== name) continue;
      if (parts[2] !== 'ack') continue;
      res.push(new Store(name, parts[1], self.keys));
    }
    return res;
  }

  each(function(store) {
    if (store.id === self.id) return;
    if (self._schedule.now() - store.get(self.keys.ACK) < self.timeouts.RECLAIM_TIMEOUT) return;
    tryReclaim(store);
  }, findOtherQueues(this.name));

  this._schedule.run(this._checkReclaim, this.timeouts.RECLAIM_TIMER);
};

Queue.prototype._reclaim = function(id) {
  var self = this;
  var other = new Store(this.name, id, this.keys);

  var our = {
    queue: this._store.get(this.keys.QUEUE) || []
  };

  var their = {
    inProgress: other.get(this.keys.IN_PROGRESS) || {},
    queue: other.get(this.keys.QUEUE) || []
  };

  // add their queue to ours, resetting run-time to immediate and copying the attempt#
  each(function(el) {
    our.queue.push({
      item: el.item,
      attemptNumber: el.attemptNumber,
      time: self._schedule.now()
    });
  }, their.queue);

  // if the queue is abandoned, all the in-progress are failed. retry them immediately and increment the attempt#
  each(function(el) {
    our.queue.push({
      item: el.item,
      attemptNumber: el.attemptNumber + 1,
      time: self._schedule.now()
    });
  }, their.inProgress);

  our.queue = our.queue.sort(function(a,b) {
    return a.time - b.time;
  });

  this._store.set(this.keys.QUEUE, our.queue);

  // remove all keys
  other.remove(this.keys.IN_PROGRESS);
  other.remove(this.keys.QUEUE);
  other.remove(this.keys.RECLAIM_START);
  other.remove(this.keys.RECLAIM_END);
  other.remove(this.keys.ACK);

  // process the new items we claimed
  this._processHead();
};

module.exports = Queue;
